### PR TITLE
#320 Added support for WAF v2 based on waf parameter

### PR DESCRIPTION
**Pull request by janario**
Related to https://github.com/zalando-incubator/kube-ingress-aws-controller/issues/320

Added support for WAF v2, based on the parameter when it is an ARN of WAF2 it will create the right template

### CODE REVIEW

File: aws/cf_template.go [hunk 1]
```diff
@@ -268,10 +274,17 @@ func generateTemplate(spec *stackSpec) (string, error) {
...
-		template.AddResource("WAFAssociation", &cloudformation.WAFRegionalWebACLAssociation{
-			ResourceArn: cloudformation.Ref("LB").String(),
-			WebACLID:    cloudformation.String(spec.wafWebAclId),
-		})
+		if strings.HasPrefix(spec.wafWebAclId, "arn:aws:wafv2") {
+			template.AddResource("WAF2WAC1KPI3", &cloudformation.WAFv2WebACLAssociation{
+				ResourceArn: cloudformation.Ref("LB").String(),
+				WebACLArn:   cloudformation.String(spec.wafWebAclId),
+			})
+		} else {
```
Comment (szuecs) [inline aws/cf_template.go:282]: What if we have no waf ID specified?
We should omit creating one as far as I understand.
Related https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/341

Comment (janario) [reply aws/cf_template.go:282]: It won't get into the first if

> if spec.loadbalancerType == LoadBalancerTypeApplication && spec.wafWebAclId != "" {

* no waf, out of the entire if
* waf arn v2, template for v2
* waf id (v1), template for v1 as it used to be

---
But depending if #341 gets merged first this one has to be changed

waf2 (I suppose this)
> WebACLArn:    cloudformation.Ref(parameterLoadBalancerWAFWebACLIdParameter).String(),

waf1
> WebACLID:    cloudformation.Ref(parameterLoadBalancerWAFWebACLIdParameter).String(),

File: aws/cf_template.go [hunk 2]
```diff
@@ -268,10 +274,17 @@ func generateTemplate(spec *stackSpec) (string, error) {
 	})
 
 	if spec.loadbalancerType == LoadBalancerTypeApplication && spec.wafWebAclId != "" {
-		template.AddResource("WAFAssociation", &cloudformation.WAFRegionalWebACLAssociation{
-			ResourceArn: cloudformation.Ref("LB").String(),
-			WebACLID:    cloudformation.String(spec.wafWebAclId),
-		})
+		if strings.HasPrefix(spec.wafWebAclId, "arn:aws:wafv2") {
+			template.AddResource("WAF2WAC1KPI3", &cloudformation.WAFv2WebACLAssociation{
```
Comment (mikkeloscar) [inline aws/cf_template.go:278]: What does this name mean? `WAF2WAC1KPI3` can we use something more human readable? :)

Comment (janario) [reply aws/cf_template.go]: Good point

I got it from the visual template, but it can be renamed

Changed it to the same as from v1 `WAFAssociation`

**Comment by mikkeloscar** [comment]
@janario If you can just fix the DCO check (signoff your commit) then it lgtm.

**Comment by janario** [comment]
> @janario If you can just fix the DCO check (signoff your commit) then it lgtm.

@mikkeloscar thanks for the reminder I forgot the signoff 

Fixed it :+1:

**Comment by mikkeloscar** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Comment by aryszka** [comment]
@janario thanks for the contribution!

File: aws/cf_template.go (hunk 1 above)
Comment (ruudk) [reply aws/cf_template.go:282]: I think this merge went wrong, see https://github.com/zalando-incubator/kube-ingress-aws-controller/issues/331#issuecomment-617031946

@janario @aryszka @mikkeloscar @szuecs

File: aws/cf_template.go [hunk 3]
```diff
@@ -82,6 +83,11 @@ func generateTemplate(spec *stackSpec) (string, error) {
 			Description: "H2 Enabled",
 			Default:     "true",
 		},
+		parameterLoadBalancerWafWebACLIdParameter: &cloudformation.Parameter{
+			Type:        "String",
+			Description: "WAF Id or ARN",
+			Default:     "",
```
Comment (szuecs) [inline aws/cf_template.go:89]: This is what I expect to be a problem mentioned in https://github.com/zalando-incubator/kube-ingress-aws-controller/issues/331
Could it be that this ends in the cf stack?

Comment (janario) [reply aws/cf_template.go:89]: I noticed the error while I was testing in my cluster

It looked like master was broken, I fixed in order to unblock my test

Sorry, I should have pointed it here

Comment (janario) [reply aws/cf_template.go:89]: I mean, it failed with an error even when I didn't provide an WAF config

**Pull request merged**